V3 error compiling any project

### Describe the bug

When attempting to build any project (even hello world) with v3 it gives the following error:

```
failed to build cached module object /tmp/v_1000/v3_module_cache_80b3d4e634fc7c30/4bee31a2e3a50d93/builtin_43995d714e877460_6d25139443b2fc25.o:
/tmp/v_1000/v3_module_cache_80b3d4e634fc7c30/4bee31a2e3a50d93/builtin_43995d714e877460_6d25139443b2fc25.c.tmp.448546.01KYA4W435WW93K1F0HKK16Y9B.c:3361:7: error: conflicting types for ‘pthread_getspecific’; have ‘void *(u64)’ {aka ‘void *(long long unsigned int)’}
 3361 | void* pthread_getspecific(u64 key);
      |       ^~~~~~~~~~~~~~~~~~~
/tmp/v_1000/v3_module_cache_80b3d4e634fc7c30/4bee31a2e3a50d93/builtin_43995d714e877460_6d25139443b2fc25.c.tmp.448546.01KYA4W435WW93K1F0HKK16Y9B.c:286:7: note: previous declaration of ‘pthread_getspecific’ with type ‘void *(pthread_key_t)’ {aka ‘void *(long unsigned int)’}
  286 | void* pthread_getspecific(pthread_key_t key);
      |       ^~~~~~~~~~~~~~~~~~~
```

### Reproduction Steps

v new testproj
cd testproj
v3 .

### Expected Behavior

It should compile

### Current Behavior

See above error

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

master

### Environment details (OS name and version, etc.)

|V full version      |V 0.5.2 f33feed16f62265f3d982344b8d19d31beee16de.c82d3f0
|:-------------------|:-------------------
|OS                  |linux, "Linux"
|Processor           |12 cpus, 64bit, little endian, AMD Ryzen 5 7640U w/ Radeon 760M Graphics
|Memory              |2.41GB/14.93GB
|                    |
|V executable        |/home/sirsegv/.v/v
|V last modified time|2026-07-24 13:26:06
|                    |
|V home dir          |OK, value: /home/sirsegv/.v
|VMODULES            |OK, value: /home/sirsegv/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/sirsegv
|                    |
|Git version         |git version 2.55.0
|V git status        |0.5.1-2297-gc82d3f08
|.git/config present |true
|                    |
|cc version          |cc (GCC) 16.1.1 20260625
|gcc version         |gcc (GCC) 16.1.1 20260625
|clang version       |clang version 22.1.8
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.43

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

Can confirm (on [currently] latest commit [`c82d3f08271e9324d6fb8de3c251e9c0e1a9154b`](https://github.com/vlang/v/commit/c82d3f08271e9324d6fb8de3c251e9c0e1a9154b)):
```console
$ v -gc none -o v3 vlib/v3/v3.v
$ v new hellow
Input your project description: 
Input your project version: (0.0.0) 
Input your project license: (MIT) 
Initialising ...
Created binary (application) project `hellow`
$ cd hellow
$ ../v3 .
=== v3 benchmark ===
  parse setup/cache       14.33 ms       44 MB RSS       44 MB peak
  parse .vh                0.00 ms       44 MB RSS       44 MB peak
  parse .v (parallel)     73.24 ms       44 MB RSS       44 MB peak
  resolve imports          8.44 ms       44 MB RSS       44 MB peak
    parsed .vh files             0 files
      .vh files: 
    parsed .vh lines             0 lines
    parsed .v files              113 files
      .v files: ~/.vlang/vlib/builtin/array.v ~/.vlang/vlib/builtin/array_notd_gcboehm_opt.v ~/.vlang/vlib/builtin/autostr.v ~/.vlang/vlib/builtin/builtin.v ~/.vlang/vlib/builtin/chan_option_result.v ~/.vlang/vlib/builtin/grapheme.v ~/.vlang/vlib/builtin/input_rune_iterator.v ~/.vlang/vlib/builtin/int.v ~/.vlang/vlib/builtin/int_notd_new_int.v ~/.vlang/vlib/builtin/map.v ~/.vlang/vlib/builtin/map_notd_gcboehm_opt.v ~/.vlang/vlib/builtin/meta_assert.v ~/.vlang/vlib/builtin/meta_attribute.v ~/.vlang/vlib/builtin/meta_enum.v ~/.vlang/vlib/builtin/meta_function.v ~/.vlang/vlib/builtin/meta_struct.v ~/.vlang/vlib/builtin/meta_variant.v ~/.vlang/vlib/builtin/ownership_interface_d_v3_backend.v ~/.vlang/vlib/builtin/ownership_notd_ownership.v ~/.vlang/vlib/builtin/reuse.v ~/.vlang/vlib/builtin/rune.v ~/.vlang/vlib/builtin/rune_map.v ~/.vlang/vlib/builtin/sorted_map.v ~/.vlang/vlib/builtin/string.v ~/.vlang/vlib/builtin/string_charptr_byteptr_helpers.v ~/.vlang/vlib/builtin/string_interpolation.v ~/.vlang/vlib/builtin/utf8.v ~/.vlang/vlib/builtin/allocation.c.v ~/.vlang/vlib/builtin/backtraces.c.v ~/.vlang/vlib/builtin/backtraces_nix.c.v ~/.vlang/vlib/builtin/builtin.c.v ~/.vlang/vlib/builtin/builtin_backtraces_nix.c.v ~/.vlang/vlib/builtin/builtin_nix.c.v ~/.vlang/vlib/builtin/builtin_notd_gcboehm.c.v ~/.vlang/vlib/builtin/builtin_notd_use_libbacktrace.c.v ~/.vlang/vlib/builtin/cfns.c.v ~/.vlang/vlib/builtin/cfns_wrapper.c.v ~/.vlang/vlib/builtin/character_inout.c.v ~/.vlang/vlib/builtin/float.c.v ~/.vlang/vlib/builtin/map.c.v ~/.vlang/vlib/builtin/option.c.v ~/.vlang/vlib/builtin/panicing.c.v ~/.vlang/vlib/builtin/prealloc.c.v ~/.vlang/vlib/builtin/printing.c.v ~/.vlang/vlib/builtin/utf8.c.v ~/.vlang/vlib/builtin/vgc_notd_vgc.c.v ./main.v ~/.vlang/vlib/strings/similarity.v ~/.vlang/vlib/strings/strings.v ~/.vlang/vlib/strings/builder.c.v ~/.vlang/vlib/strings/strings.c.v ~/.vlang/vlib/strconv/atoi.v ~/.vlang/vlib/strconv/atou.v ~/.vlang/vlib/strconv/f64_str.v ~/.vlang/vlib/strconv/format.v ~/.vlang/vlib/strconv/structs.v ~/.vlang/vlib/strconv/tables.v ~/.vlang/vlib/strconv/utilities.v ~/.vlang/vlib/strconv/write_dec.v ~/.vlang/vlib/strconv/atof.c.v ~/.vlang/vlib/strconv/atofq.c.v ~/.vlang/vlib/strconv/f32_str.c.v ~/.vlang/vlib/strconv/f64_str.c.v ~/.vlang/vlib/strconv/format_mem.c.v ~/.vlang/vlib/strconv/ftoa.c.v ~/.vlang/vlib/strconv/number_to_base.c.v ~/.vlang/vlib/strconv/utilities.c.v ~/.vlang/vlib/strconv/vprintf.c.v ~/.vlang/vlib/sync/common_mutex.v ~/.vlang/vlib/sync/cond.v ~/.vlang/vlib/sync/many_times.v ~/.vlang/vlib/sync/once.v ~/.vlang/vlib/sync/tls.v ~/.vlang/vlib/sync/channels.c.v ~/.vlang/vlib/sync/sync.c.v ~/.vlang/vlib/sync/sync_default.c.v ~/.vlang/vlib/sync/thread_default.c.v ~/.vlang/vlib/sync/tls_default.c.v ~/.vlang/vlib/sync/waitgroup.c.v ~/.vlang/vlib/hash/hash.v ~/.vlang/vlib/hash/wyhash.v ~/.vlang/vlib/hash/wyhash.c.v ~/.vlang/vlib/math/bits/bits.amd64.v ~/.vlang/vlib/math/bits/bits.v ~/.vlang/vlib/math/bits/bits_tables.v ~/.vlang/vlib/math/bits/unsafe_bits.v ~/.vlang/vlib/math/bits/bits.c.v ~/.vlang/vlib/sync/stdatomic/1.declarations.c.v ~/.vlang/vlib/sync/stdatomic/atomic.c.v ~/.vlang/vlib/time/chrono.v ~/.vlang/vlib/time/date_time_parser.v ~/.vlang/vlib/time/duration.v ~/.vlang/vlib/time/format.v ~/.vlang/vlib/time/operator.v ~/.vlang/vlib/time/parse.v ~/.vlang/vlib/time/stopwatch.v ~/.vlang/vlib/time/time.v ~/.vlang/vlib/time/unix.v ~/.vlang/vlib/time/chrono.c.v ~/.vlang/vlib/time/json_custom.c.v ~/.vlang/vlib/time/parse.c.v ~/.vlang/vlib/time/time.c.v ~/.vlang/vlib/time/time_linux.c.v ~/.vlang/vlib/time/time_nix.c.v ~/.vlang/vlib/time/timerfd.c.v ~/.vlang/vlib/rand/mini_math.v ~/.vlang/vlib/rand/rand.v ~/.vlang/vlib/rand/rand.c.v ~/.vlang/vlib/rand/config/config.v ~/.vlang/vlib/rand/wyrand/wyrand.v ~/.vlang/vlib/rand/wyrand/z_wyrand.c.v ~/.vlang/vlib/rand/seed/seed.v ~/.vlang/vlib/rand/buffer/buffer.v
    parsed .v lines              28014 lines
  cache lookup             4.21 ms       45 MB RSS       45 MB peak
  check (parallel)       340.65 ms      210 MB RSS      210 MB peak
  markused               200.29 ms      242 MB RSS      242 MB peak
  transform (parallel)   222.33 ms      372 MB RSS      372 MB peak
  annotate types         124.84 ms      396 MB RSS      396 MB peak
  monomorphize           168.12 ms      418 MB RSS      418 MB peak
  cgen (parallel)        211.43 ms      484 MB RSS      487 MB peak
  C object cache          17.89 ms      484 MB RSS      487 MB peak
failed to build cached module object /tmp/v_1000/v3_module_cache_9cf5f22029779189/138194e9394b31f8/builtin_39b78619b7058993_708daef884d94fce.o:
/tmp/v_1000/v3_module_cache_9cf5f22029779189/138194e9394b31f8/builtin_39b78619b7058993_708daef884d94fce.c.tmp.3636183.01KYAAKGBJR4GYYTE0FYY5TE2V.c:3361:7: error: conflicting types for ‘pthread_getspecific’; have ‘void *(u64)’ {aka ‘void *(long long unsigned int)’}
 3361 | void* pthread_getspecific(u64 key);
      |       ^~~~~~~~~~~~~~~~~~~
/tmp/v_1000/v3_module_cache_9cf5f22029779189/138194e9394b31f8/builtin_39b78619b7058993_708daef884d94fce.c.tmp.3636183.01KYAAKGBJR4GYYTE0FYY5TE2V.c:286:7: note: previous declaration of ‘pthread_getspecific’ with type ‘void *(pthread_key_t)’ {aka ‘void *(long unsigned int)’}
  286 | void* pthread_getspecific(pthread_key_t key);
      |       ^~~~~~~~~~~~~~~~~~~
/tmp/v_1000/v3_module_cache_9cf5f22029779189/138194e9394b31f8/builtin_39b78619b7058993_708daef884d94fce.c.tmp.3636183.01KYAAKGBJR4GYYTE0FYY5TE2V.c:3365:5: error: conflicting types for ‘pthread_setspecific’; have ‘i32(u64,  const void *)’ {aka ‘int(long long unsigned int,  const void *)’}
 3365 | i32 pthread_setspecific(u64 key, const void* const_ptr);
      |     ^~~~~~~~~~~~~~~~~~~
/tmp/v_1000/v3_module_cache_9cf5f22029779189/138194e9394b31f8/builtin_39b78619b7058993_708daef884d94fce.c.tmp.3636183.01KYAAKGBJR4GYYTE0FYY5TE2V.c:287:5: note: previous declaration of ‘pthread_setspecific’ with type ‘int(pthread_key_t,  const void *)’ {aka ‘int(long unsigned int,  const void *)’}
  287 | int pthread_setspecific(pthread_key_t key, const void* const_ptr);
      |     ^~~~~~~~~~~~~~~~~~~

```
Also v3 can't compile itself via:
```
./v3 -o v3new vlib/v3/v3.v
```
<details><summary>v doctor</summary>
<p>

|V full version      |V 0.5.2 2f3c0f1f397c08bfc5ef73e7ff62794471b2b811.c82d3f0
|:-------------------|:-------------------
|OS                  |linux, Linux version 7.0.12-arch1-1 (linux@archlinux) (gcc (GCC) 16.1.1 20260430, GNU ld (GNU Binutils) 2.46.0) #1 SMP PREEMPT_DYNAMIC Wed, 10 Jun 2026 08:58:02 +0000
|Processor           |16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P
|Memory              |3.35GB/15.34GB
|                    |
|V executable        |/home/ge/.vlang/v
|V last modified time|2026-07-24 14:59:17
|                    |
|V home dir          |OK, value: /home/ge/.vlang
|VMODULES            |OK, value: /home/ge/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/ge/.vlang
|                    |
|Git version         |git version 2.54.0
|V git status        |weekly.2026.03-2671-gc82d3f08
|.git/config present |true
|                    |
|cc version          |cc (GCC) 16.1.1 20260430
|gcc version         |gcc (GCC) 16.1.1 20260430
|clang version       |clang version 22.1.6
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.43


</p>
</details>

--- a/vlib/v3/gen/c/fn.v
+++ b/vlib/v3/gen/c/fn.v
@@ -12946,11 +12946,14 @@ const c_preamble_declared_extern_symbols = {
 	'pthread_cond_wait':             true
 	'pthread_create':                true
 	'pthread_detach':                true
+	'pthread_getspecific':           true
 	'pthread_join':                  true
+	'pthread_key_create':            true
 	'pthread_mutex_destroy':         true
 	'pthread_mutex_init':            true
 	'pthread_mutex_lock':            true
 	'pthread_mutex_unlock':          true
+	'pthread_setspecific':           true
 	'malloc':                        true
 	'calloc':                        true
 	'realloc':                       true

--- a/vlib/v3/gen/c/source_directive_test.v
+++ b/vlib/v3/gen/c/source_directive_test.v
@@ -68,6 +68,24 @@ fn test_cache_extern_declaration_avoids_tgmath_macro_expansion() {
 	assert c_macro_safe_extern_decl('custom', 'int custom(int x);') == 'int custom(int x);'
 }
 
+fn test_cache_extern_filter_uses_pthread_preamble_declarations() {
+	mut preamble_gen := FlatGen.new()
+	preamble_gen.preamble()
+	preamble := preamble_gen.sb.str()
+	assert preamble.contains('int pthread_key_create(pthread_key_t* key, void (*dtor)(void*));')
+	assert preamble.contains('void* pthread_getspecific(pthread_key_t key);')
+	assert preamble.contains('int pthread_setspecific(pthread_key_t key, const void* const_ptr);')
+	assert !preamble.contains('pthread_key_delete(')
+
+	mut g := FlatGen.new()
+	g.set_cache_split(true)
+
+	assert !g.should_emit_c_extern_decl('pthread_key_create')
+	assert !g.should_emit_c_extern_decl('pthread_getspecific')
+	assert !g.should_emit_c_extern_decl('pthread_setspecific')
+	assert g.should_emit_c_extern_decl('pthread_key_delete')
+}
+
 fn test_builtin_abi_compat_macros_precede_late_c_source() {
 	mut g := FlatGen.new()
 	g.has_builtins = true

--- a/vlib/v3/tests/module_cache_test.v
+++ b/vlib/v3/tests/module_cache_test.v
@@ -56,6 +56,45 @@ fn run_module_cache_binary(path string) string {
 	return result.output.trim_space()
 }
 
+fn test_cached_sync_module_uses_preamble_pthread_declarations() {
+	v3_bin := build_module_cache_v3()
+	root := os.join_path(os.temp_dir(), 'v3_module_cache_sync_pthread_${os.getpid()}')
+	os.rmdir_all(root) or {}
+	os.mkdir_all(root) or { panic(err) }
+	defer {
+		os.rmdir_all(root) or {}
+	}
+	main_file := os.join_path(root, 'main.v')
+	write_module_cache_file(root, 'main.v', 'module main
+
+import sync
+
+fn main() {
+	mut m := sync.new_mutex()
+	m.lock()
+	m.unlock()
+	m.destroy()
+	println("V3_CACHE_SYNC_OK")
+}
+')
+	cache_dir := os.join_path(root, 'cache')
+	first_output := os.join_path(root, 'first')
+	first :=
+		os.execute('V3CACHE=${os.quoted_path(cache_dir)} ${os.quoted_path(v3_bin)} -o ${os.quoted_path(first_output)} ${os.quoted_path(main_file)}')
+	assert first.exit_code == 0, first.output
+	assert !first.output.contains('check (cached)'), first.output
+	assert run_module_cache_binary(first_output) == 'V3_CACHE_SYNC_OK'
+
+	second_output := os.join_path(root, 'second')
+	second :=
+		os.execute('V3CACHE=${os.quoted_path(cache_dir)} ${os.quoted_path(v3_bin)} -o ${os.quoted_path(second_output)} ${os.quoted_path(main_file)}')
+	assert second.exit_code == 0, second.output
+	assert second.output.contains('check (cached)'), second.output
+	assert second.output.contains('C module plan (cached)'), second.output
+	assert second.output.contains('cgen (cached)'), second.output
+	assert run_module_cache_binary(second_output) == 'V3_CACHE_SYNC_OK'
+}
+
 fn test_cached_header_preserves_mutable_pointer_parameters() {
 	v3_bin := build_module_cache_v3()
 	root := os.join_path(os.temp_dir(), 'v3_cached_mut_pointer_param_${os.getpid()}')

--- a/vlib/v3/tests/parallel_cgen_test.v
+++ b/vlib/v3/tests/parallel_cgen_test.v
@@ -369,23 +369,32 @@ fn test_prealloc_keeps_parallel_transform_enabled() {
 	assert compile.output.contains('cgen (parallel)'), compile.output
 }
 
-fn test_parallel_transform_selfhost_builds_v3() {
+fn test_parallel_transform_generates_v3_c_with_vjobs_4_and_12() {
 	v3_bin := build_parallel_prod_v3()
-	bin_out := os.join_path(os.temp_dir(), 'v3_parallel_selfhost_out_${os.getpid()}')
-	os.rm(bin_out) or {}
-	os.rm(bin_out + '.c') or {}
-	compile := os.execute('VJOBS=2 ${v3_bin} -nocache -building-v -o ${bin_out} ${parallel_v3_src}')
-	assert compile.exit_code == 0, compile.output
-	assert compile.output.contains('transform'), compile.output
-	assert compile.output.contains('cgen (parallel)'), compile.output
-	assert os.exists(bin_out), compile.output
-	c_code := os.read_file(bin_out + '.c') or { panic(err) }
+	c_out_4 := os.join_path(os.temp_dir(), 'v3_parallel_selfhost_out_4_${os.getpid()}.c')
+	os.rm(c_out_4) or {}
+	cgen_4 :=
+		os.execute('VJOBS=4 ${v3_bin} -nocache -building-v -b c -o ${c_out_4} ${parallel_v3_src}')
+	assert cgen_4.exit_code == 0, cgen_4.output
+	assert cgen_4.output.contains('transform (parallel)'), cgen_4.output
+	assert cgen_4.output.contains('cgen (parallel)'), cgen_4.output
+	assert os.exists(c_out_4), cgen_4.output
+	c_code := os.read_file(c_out_4) or { panic(err) }
 	assert c_code.contains('Array_u8__bytestr'), c_code
 	assert c_code.contains('Array_u8__hex'), c_code
 	assert c_code.contains('typedef void* pthread_t;'), c_code
 	assert !c_code.contains('typedef struct pthread_t pthread_t;'), c_code
 	assert c_code.contains('flat_cgen_chunk_thread'), c_code
 	assert c_code.contains('run_parallel_transform'), c_code
+
+	c_out_12 := os.join_path(os.temp_dir(), 'v3_parallel_selfhost_out_12_${os.getpid()}.c')
+	os.rm(c_out_12) or {}
+	cgen_12 :=
+		os.execute('VJOBS=12 ${v3_bin} -nocache -building-v -b c -o ${c_out_12} ${parallel_v3_src}')
+	assert cgen_12.exit_code == 0, cgen_12.output
+	assert cgen_12.output.contains('transform (parallel)'), cgen_12.output
+	assert cgen_12.output.contains('cgen (parallel)'), cgen_12.output
+	assert os.exists(c_out_12), cgen_12.output
 }
 
 fn test_no_parallel_directory_selfhost_omits_parallel_support() {

--- a/vlib/v3/transform/transform_parallel_notd_v3_no_parallel.v
+++ b/vlib/v3/transform/transform_parallel_notd_v3_no_parallel.v
@@ -18,14 +18,6 @@ const max_parallel_transform_jobs = 7
 // append into pre-partitioned capacity regions, so extra threads cost no
 // clone memory; cap by core count only.
 const max_shared_transform_jobs = 10
-// Split the shared-base transform into more chunks than pool threads: the
-// persistent pool hands queued chunks to whichever worker frees up first, so
-// finer chunks level the load across asymmetric (performance/efficiency)
-// cores while the fixed region merge order keeps node ids deterministic.
-const shared_transform_chunk_factor = 3
-// Keep enough items per chunk that per-chunk append-region headroom (sized
-// proportionally to estimated cost) still averages out expansion variance.
-const min_shared_transform_chunk_items = 24
 const max_parallel_monomorph_jobs = 10
 const scoped_transform_worker_batches = 1
 const scoped_transform_master_batches = 1
@@ -1373,21 +1365,9 @@ fn (mut t Transformer) run_parallel_transform_shared(items []FnWorkItem, base_no
 		return false
 	} $else {
 		t.tc.freeze_type_cache_for_forks()
-		mut chunk_target := n_jobs * shared_transform_chunk_factor
-		max_chunks_by_items := items.len / min_shared_transform_chunk_items
-		if chunk_target > max_chunks_by_items {
-			chunk_target = max_chunks_by_items
-		}
-		if chunk_target < n_jobs {
-			chunk_target = n_jobs
-		}
-		mut chunks := split_work_items_ex(items, chunk_target, false)
+		mut chunks := split_work_items_ex(items, n_jobs, false)
 		chunk_count := chunks.len
 		thread_count := chunk_count - 1
-		// The master transforms chunk 0 in place and then this many additional
-		// forked chunks synchronously (keeping its former ~1/n_jobs share),
-		// while the pool workers drain the rest of the queue dynamically.
-		master_sync_chunks := chunk_count / n_jobs - 1
 		// Partition the reserved capacity into per-chunk append regions,
 		// proportional to chunk cost (the caller reserved ~2x the expected
 		// total growth for this pool).
@@ -1456,8 +1436,7 @@ fn (mut t Transformer) run_parallel_transform_shared(items []FnWorkItem, base_no
 			tasks << workers.Task{
 				run:        shared_chunk_thread
 				arg:        unsafe { voidptr(&args[ci]) }
-				force_sync: ci <= master_sync_chunks || fail == 'transform:all'
-					|| fail == 'transform:${helper_idx}'
+				force_sync: ci == 0 || fail == 'transform:all' || fail == 'transform:${helper_idx}'
 			}
 		}
 		transform_worker_scope_leave(setup_scope)

